### PR TITLE
Support getting snapshot at or right before the given timestamp

### PR DESCRIPTION
**Pull request by chinmay-bhat**
Bring support to retrieve a snapshot before a particular timestamp, which is needed to perform Spark procedure like [rollback_to_timestamp](https://iceberg.apache.org/docs/1.5.1/spark-procedures/#rollback_to_timestamp). 

- [x] create ancestors_of() ([relevant spark procedure](https://iceberg.apache.org/docs/nightly/spark-procedures/?h=create_changelog_view#ancestors_of))
- [x] create snapshot_as_of_timestamp()
- [x] add tests

See [comment in issue](https://github.com/apache/iceberg-python/issues/737#issuecomment-2113105812)

### CODE REVIEW

File: pyiceberg/table/metadata.py
Comment (chinmay-bhat) [inline pyiceberg/table/metadata.py:236]: Instead of searching for a valid snapshot in self.snapshots (which has all valid snapshots irrespective of branch), I should probably write an API to find all ancestors of current snapshot and iterate over those snapshots only.

**Comment by ndrluis** [comment]
Hello @chinmay-bhat,

I noticed that you are implementing the ancestors_of method, and we have another pull request (#533) that is implementing the same behavior in another place as a function with a different output (Iterable[Snapshot] instead of a List[tuple[int, int]]) and signature (expects a Snapshot instead of a Snapshot ID).

I believe that we need to discuss and choose which one we want to have in the codebase.

cc/ @HonahX @Fokko @syun64

**Comment by syun64** [comment]
Hi @ndrluis thank you for flagging this! That PR went under my radar, and I'm excited to see a changelog scanning feature being implemented already on PyIceberg.

As for the question on the output type, I'm +1 for using Iterable[Snapshot] because I have a preference for using a class with set attributes than using a tuple.

Im also +1 for introducing the feature in this separate PR, since it's a much simpler feature in itself we can introduce quickly. WDYT?

**Comment by chinmay-bhat** [comment]
Happy to update the output type to Iterable[Snapshot]! Also I really like how concise the ancestors_of function is in the other PR.

File: pyiceberg/table/__init__.py
Comment (HonahX) [inline pyiceberg/table/__init__.py:1299]:
```suggestion
                if snapshot and prev_timestamp < snapshot.timestamp_ms <= timestamp_ms:
```
I think we could also include the given timestamp, making the method return the latest snapshot up until the given timestamp.

Also, how about implementing this by iterating over `table.history()`? The `snapshot_log` field in metadata contains a list of `snapshot_id` + `timestamp` pair so we do not need to re-generate the ancestors for current snapshot.

**Review by HonahX — COMMENTED** [review]
@chinmay-bhat Thanks for working on this! I left some comments. Love to see more snapshot utils being added.

File: pyiceberg/table/__init__.py
Comment (chinmay-bhat) [reply pyiceberg/table/__init__.py:1299]: > I think we could also include the given timestamp, making the method return the latest snapshot up until the given timestamp. 

That was my initial approach, however, I saw the [java implementation](https://github.com/apache/iceberg/blob/795fea9446723cf9c2dbc3cac07883cace8997c9/core/src/main/java/org/apache/iceberg/SetSnapshotOperation.java#L146) for rollback which is not inclusive of the given timestamp. This [implementation](https://github.com/apache/iceberg/blob/795fea9446723cf9c2dbc3cac07883cace8997c9/core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java#L350), which is what you describe, exists but is not used for rollback. So I thought there must be a design decision I am unaware of to keep it not inclusive.
WDYT - should I still change it to be inclusive of the timestamp?

> Also, how about implementing this by iterating over table.history()? The snapshot_log field in metadata contains a list of snapshot_id + timestamp pair so we do not need to re-generate the ancestors for current snapshot.

Ok!

File: pyiceberg/table/__init__.py
Comment (chinmay-bhat) [reply pyiceberg/table/__init__.py:1303]: > Is there any use-case for this API that we may prefer a list?
When we want to retrieve the snapshot_ids of ancestors, although the list is not needed, we can use Iterable too.

File: pyiceberg/table/snapshots.py
Comment (chinmay-bhat) [reply pyiceberg/table/snapshots.py:428]: good catch, I missed this

File: pyiceberg/table/__init__.py
Comment (HonahX) [reply pyiceberg/table/__init__.py]: That's an interesting finding. I am also not sure the exact decision behind whether inclusive. Although whether to include the exact timestamp does not matter a lot in this use case, I think we may still want to keep the edge case behavior the same as Java: that we do not include the given timestamp when `rollback_to_timestamp`. 

Since this will be a public util method, how about adding another parameter like

- `inclusive`
- `include_given_timstamp`
- ...
that default to `True` and we disable it when using it in `rollback_to_timestamp` implementation?

File: pyiceberg/table/__init__.py
Comment (HonahX) [inline pyiceberg/table/__init__.py:1303]:
```suggestion
        return ancestors_of(self.current_snapshot(), self.metadata)
```
I think we do not need this any more?

Comment (chinmay-bhat) [reply pyiceberg/table/__init__.py]: I've resolved the ignore error.

I think we don't need `current_ancestors()` anymore. I'm not using it in this PR, nor do we need this as a public method, since table.history() does a very similar job. 

The `ancestors_of()` is no longer being used, but as mentioned in previous discussions in this PR, it is a small function that is used in the cherrypick PR, once these changes are merged.

File: pyiceberg/table/__init__.py
Comment (chinmay-bhat) [reply pyiceberg/table/__init__.py]: updated to use inclusive parameter!

File: pyiceberg/table/__init__.py
Comment (HonahX) [reply pyiceberg/table/__init__.py]: +1 on removing `current_ancestors()` if we do not need it. I think we can keep `ancestor_of()` since it will be an useful helper. We just need to make sure there are tests for it. (Which you have already added for `current_ancestors()`:raised_hands:).

Comment (Fokko) [reply pyiceberg/table/__init__.py]: Thanks for working on this @chinmay-bhat, I also [ran into the confusion around `current-ancestor` myself](https://github.com/apache/iceberg/issues/6930). Thanks for adding that to the docs itself.

**Review by HonahX — APPROVED** [review]
Thanks for updating! I think this is good to go when the lint issue is fixed.

**Review by Fokko — APPROVED** [review]
Looking great @chinmay-bhat and thanks for the review @HonahX 🙌

**Comment by chinmay-bhat** [comment]
Thank you for the review @Fokko and @HonahX ! 🚀

**Comment by HonahX** [comment]
Merged! Thanks @chinmay-bhat for the great work! Thanks @Fokko @syun64 @ndrluis for the review and discussions!

**Comment by syun64** [comment]
Congrats on your first PR @chinmay-bhat !

**Pull request merged**
